File: javascript/luhn/luhn.js
function Luhn(input) {
	this.number = input.toString();
	
	this.checkDigit = parseInt(this.number[this.number.length - 1]);
	this.addends = this.computeAddends();
	this.checksum = this.computeChecksum();
	this.valid = this.isValid();
}

Luhn.prototype.computeAddends = function() {
	var addendsArray = [];
	
	var numLenght = this.number.length;
	for (var i = numLenght - 1; i >= 0; --i) {
		if (0 === (numLenght - i) % 2) {
			var doubledNumber = parseInt(this.number[i]) * 2; 
			addendsArray.unshift((doubledNumber > 9) ? doubledNumber - 9 : doubledNumber);
		} else {
			addendsArray.unshift(parseInt(this.number[i]));
		}
	}
	
	return addendsArray;
};

Luhn.prototype.computeChecksum = function() {
	var computedChecksum = 0;
	
	var addendsLength = this.addends.length;
	for (var i = 0; i < addendsLength; ++i) {
		computedChecksum += this.addends[i];
	}
	
	return computedChecksum;
};

Luhn.prototype.isValid = function() {
	return (0 === (this.checksum % 10)) ? true : false;
};

// this is a static method
Luhn.create = function(input) {
	// For easily dealing with digits. Not the most optimized idea here, though.
	input = input.toString();
	var paddedInput = parseInt(input + '0');
	var number = new Luhn(paddedInput);

	if (!number.isValid()) {	
		var numChecksum = number.checksum.toString();
		var generatedCheckDigit = 10 - parseInt(numChecksum[numChecksum.length - 1]);
		
		return parseInt(input + generatedCheckDigit);
	}
	
	return paddedInput;
};

module.exports = Luhn;

File: python/luhn/luhn.py
class Luhn:
    def __init__(self, num):
        self._number = str(num)
    
    def addends(self):
        digits = self._number
        addends_arr = []
    
        num_len = len(digits)
        for i in range(num_len - 1, -1, -1):
            if (0 == (num_len - i) % 2):
                doubled_num = int(digits[i]) * 2
                addends_arr.append((doubled_num - 9) if (doubled_num > 9) else doubled_num)
            else:
                addends_arr.append(int(digits[i]))
        
        return addends_arr
    
    def checksum(self):
        return sum(self.addends())
    
    def is_valid(self):
        return 0 == (self.checksum() % 10)
    
    # Theoretically, this is the way to create static methods. (without the `self`)
    @staticmethod
    def create(num):
        num = str(num)
        padded_input = int(num + '0')
        number = Luhn(padded_input)

        if not number.is_valid():
            num_checksum = str(number.checksum())
            generated_check_digit = 10 - int(num_checksum[len(num_checksum) - 1])
            return int(num + str(generated_check_digit))
        
        return padded_input